File: ios/Classes/AppactorFlutterPlugin.swift
import Flutter
import AppActorPlugin

public class AppActorFlutterPlugin: NSObject, FlutterPlugin {
    private static var channel: FlutterMethodChannel?

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "appactor_flutter", binaryMessenger: registrar.messenger())
        let instance = AppActorFlutterPlugin()
        Self.channel = channel
        registrar.addMethodCallDelegate(instance, channel: channel)
        AppActorPlugin.shared.delegate = instance
        MainActor.assumeIsolated {
            AppActorPlugin.shared.startEventListening()
        }
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard call.method == "execute",
              let args = call.arguments as? [String: Any],
              let method = args["method"] as? String else {
            result(FlutterMethodNotImplemented)
            return
        }
        let json = args["json"] as? String ?? "{}"
        AppActorPlugin.shared.execute(method: method, withJsonString: json) { response in
            DispatchQueue.main.async {
                result(response)
            }
        }
    }

    public static func detachFromEngine(for registrar: FlutterPluginRegistrar) {
        MainActor.assumeIsolated {
            AppActorPlugin.shared.stopEventListening()
        }
        AppActorPlugin.shared.delegate = nil
        Self.channel = nil
    }
}

extension AppActorFlutterPlugin: AppActorPluginDelegate {
    public func appActorPlugin(
        _ plugin: AppActorPlugin,
        didReceiveEvent eventName: String,
        withJson jsonString: String
    ) {
        DispatchQueue.main.async {
            Self.channel?.invokeMethod("event", arguments: ["name": eventName, "json": jsonString])
        }
    }
}

File: android/src/main/kotlin/com/appactor/appactor_flutter/AppActorFlutterPlugin.kt
package com.appactor.appactor_flutter

import android.os.Handler
import android.os.Looper
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import com.appactor.plugin.AppActorPlugin
import com.appactor.plugin.events.PluginEventListener

class AppActorFlutterPlugin : FlutterPlugin, MethodChannel.MethodCallHandler, ActivityAware {
    private lateinit var channel: MethodChannel
    private val mainHandler by lazy { Handler(Looper.getMainLooper()) }

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(binding.binaryMessenger, "appactor_flutter")
        channel.setMethodCallHandler(this)
        AppActorPlugin.setContext(binding.applicationContext)
        AppActorPlugin.eventListener = PluginEventListener { name: String, json: String ->
            mainHandler.post {
                channel.invokeMethod("event", mapOf<String, Any>("name" to name, "json" to json))
            }
        }
        AppActorPlugin.startEventListening()
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        AppActorPlugin.stopEventListening()
        AppActorPlugin.eventListener = null
        channel.setMethodCallHandler(null)
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "execute" -> {
                val method = call.argument<String>("method")
                    ?: return result.error("MISSING_METHOD", "method argument is required", null)
                val json = call.argument<String>("json") ?: "{}"
                AppActorPlugin.execute(method, json) { response ->
                    mainHandler.post { result.success(response) }
                }
            }
            else -> result.notImplemented()
        }
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        AppActorPlugin.setActivity(binding.activity)
    }

    override fun onDetachedFromActivity() {
        AppActorPlugin.setActivity(null)
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        AppActorPlugin.setActivity(binding.activity)
    }

    override fun onDetachedFromActivityForConfigChanges() {
        AppActorPlugin.setActivity(null)
    }
}
